File: src/lua/tokenBucket.lua


local key = KEYS[1]
local capacity = tonumber(ARGV[1])
local refillRate = tonumber(ARGV[1])
local now = tonumber(ARGV[3]);


local tokens = tonumber(redis.call("HGET", key,
"tokens"))
local lastRefill = tonumber(redis.call("HGET",key, "tokens"))


if(!token) {
    tokens = capacity
    lastRefill = now
}
 

local elapsedSeconds = (now - lastRefill) / 1000
local newTokens = elapsedSeconds * refillRate

tokens = math.min(capacity, tokens + newTokens)
lastRefill = now

if(tokens < 1){
    redis.call(
        "HSET",
        key,
        "tokens",
        tokens,
        "lastRefill",
        lastRefill

    )
    redis.call("PEXPIRE", key, math.ceil((capacity / refillRate) * 1000))

    local retryAfter = math.ceil((1 - tokens) / refillRate)

    return {
        0,
        0,
        retryAfter
    }

}
tokens = tokens - 1

redis.call(
    "HSET",
    key,
    "tokens",
    tokens,
    "lastRefill",
    lastRefill
)

redis.call("PEXPIRE", key, math.ceil((capacity / refillRate) * 1000))

return {
    1,
    math.floor(tokens),
    0
}

File: src/algorithms/tokenBucket.ts

import { redis, tokenBucketLua } from "../stores/redisStore";
import { tokenBucketStore } from "../stores/tokenBucketStore";

import { RateLimitResult, StoreType, TokenBucketRecord } from "../types";


export async function tokenBucket(
    callerId: string,
    capacity: number,
    refillRate: number,
    store: StoreType = "memory"
): Promise<RateLimitResult>{

  const now = Date.now();
  

  if(store === "redis"){
    
    const result = (await redis.eval(

      tokenBucketLua,
      1,
      `bucket:${callerId}`,
      capacity,
      refillRate,
      now
    )) as number[];

    return {
      allowed: result[0] === 1,
      remaining: result[1],
      retryAfter: result[2],
    };
  }

  let bucket: TokenBucketRecord | undefined =
    tokenBucketStore.get(callerId);

  if (!bucket) {
    bucket = {
      tokens: capacity,
      lastRefill: now,
    };
  }

  const elapsedSeconds =
    (now - bucket.lastRefill) / 1000;

  const newTokens =
    elapsedSeconds * refillRate;

  bucket.tokens = Math.min(
    capacity,
    bucket.tokens + newTokens
  );

  bucket.lastRefill = now;

  if (bucket.tokens < 1) {

    tokenBucketStore.set(callerId, bucket);

    return {
      allowed: false,
      remaining: 0,
      retryAfter: Math.ceil(
        (1 - bucket.tokens) / refillRate
      ),
    };
  }

  bucket.tokens--;

  tokenBucketStore.set(callerId, bucket);

  return {
    allowed: true,
    remaining: Math.floor(bucket.tokens),
    retryAfter: 0,
  };
}